src/engine/sequencer.js

```diff
@@ -168,109 +167,128 @@ class Sequencer {
-                return;
-            }
-            // If no control flow has happened, switch to next block.
-            if (thread.peekStack() === currentBlockId) {
-                thread.goToNextBlock();
-            }
-            // If no next block has been found at this point, look on the stack.
-            while (!thread.peekStack()) {
-                thread.popStack();
 
-                if (thread.stack.length === 0) {
-                    // No more stack to run!
-                    thread.status = Thread.STATUS_DONE;
-                    return;
+            if (thread.status === Thread.STATUS_RUNNING) {
+                // If no control flow has happened, switch to next block.
+                if (thread.peekStack() === currentBlockId) {
+                    thread.goToNextBlock();
                 }
-
-                const stackFrame = thread.peekStackFrame();
-                isWarpMode = stackFrame.warpMode;
-
-                if (stackFrame.isLoop) {
-                    // The current level of the stack is marked as a loop.
-                    // Return to yield for the frame/tick in general.
-                    // Unless we're in warp mode - then only return if the
-                    // warp timer is up.
-                    if (!isWarpMode ||
-                        thread.warpTimer.timeElapsed() > Sequencer.WARP_TIME) {
-                        // Don't do anything to the stack, since loops need
-                        // to be re-executed.
-                        return;
-                    }
-                    // Don't go to the next block for this level of the stack,
-                    // since loops need to be re-executed.
-                    continue;
-
-                } else if (stackFrame.waitingReporter) {
-                    // This level of the stack was waiting for a value.
-                    // This means a reporter has just returned - so don't go
-                    // to the next block for this level of the stack.
-                    return;
+                // A empty branch was pushed or we stepped out of the last block
+                // in a stack.
+                if (thread.peekStack() === null) {
+                    // Unwrap the stack until we find a loop block, a non-null
+                    // block, or the top of the stack.
+                    this.unwrapStack(thread);
                 }
-                // Get next block of existing block on the stack.
-                thread.goToNextBlock();
             }
+
+            if (thread.status === Thread.STATUS_YIELD &&
+                thread.peekStackFrame().warpMode &&
+                thread.warpTimer.timeElapsed() <= Sequencer.WARP_TIME) {
+                // In warp mode, yielded blocks are re-executed immediately.
+                thread.status = Thread.STATUS_RUNNING;
+            }
+        }
+        // Indicate the block that just executed.
+        thread.blockGlowInFrame = currentBlockId;
```

Comment:
Shouldn't this happen inside the while loop?

Reply:
Hmm. `blockGlowInFrame` is one of those thread parts that I thought was odd to be in sequencer. I think it'd make the most sense being next to wherever requestScriptGlowInFrame is set.

Looking back at this, I've done just that. I moved blockGlowInFrame into execute alongside requestScript before the loops in execute(). That'll update the blockGlowInFrame every time we run a block.